RFC-9449: DPoP

zachswasey opened the pull request:
This is a WIP to request feedback early on in the approach before I clean things up, write tests, and submit an official PR.

This implements [RFC-9449: OAuth 2.0 Demonstrating Proof of Possession (DPoP)](https://datatracker.ietf.org/doc/html/rfc9449)

I'm supporting ATProto (Bluesky) in my application, and their implementation requires DPoP and PAR (which I also have a change I'm testing that I'll put up as a WIP PR later). You can read more about their specific requirements [here](https://docs.bsky.app/docs/advanced-guides/oauth-client) for context.

Please ignore the requests integration at the moment, as I've only focused on HTTPX in FastAPI at the moment. I'll update the other integrations once the approach is more finalized.

To use:
* A dev would instantiate a rfc9449.DPoPProof, and pass that to the OAuthClient.
* The session JWK is created internally, and passed back on the token to be persisted along with the token since it needs to remain the same for the lifespan of the auth session and used for all interactions after /authorize or /par.
* A JWK can also be passed initially if desired, taken off the token that's passed into the client (if available), or have the mechanism to generate the JWK overridden if something besides ES256 is needed.
* Nonces can be returned and persisted by passing an update_nonces function as well via `DPoPProof(update_nonces=lambda token, nonces: ...)`, though this is a minor optimization solely for the first requests in a session because if an initial nonce isn't provided 9449 requires that the request should be retried and a new nonce will be returned and stored in memory, causing further requests to use that.
* TokenAuth/ClientAuth support retrying the request when a new nonce is returned from the auth/resource servers.

I explored refactoring TokenAuth/ClientAuth to support multiple Auths (separating Authorization header/client_secret_* modifications, and the DPoP proof header modifications), but that turned out to be more complicated than I originally expected. This would've introduced a new CompositeAuth to handle the pairs: TokenAuth/DPoPAuth, ClientAuth/DPoPAuth. I'm not against further exploring this idea, if it's desired by the maintainers. For now, I've implemented DPoP Proof directly in TokenAuth/ClientAuth implementations.

Feel free to give any comments or ask questions as necessary about the implementation. Once I've got go ahead for the approach I'll clean up, implement the further integrations, add tests, and put up a final PR.

Thanks!

zachswasey commented:
Whoops, I wasn't aware that a Draft PR wouldn't notify anybody to actually take a look. So, marked as Ready for Review, but again, still just WIP and looking for high level feedback before finalizing. Thanks!

azmeuk commented:
Hi. Thank you for your contribution. We had the notification even when the PR was in draft.
I'll try to review your work, but as I did not study the DPoP spec yet, it may take some days or weeks.

zachswasey commented on authlib/oauth2/rfc6749/grants/refresh_token.py:
```diff
@@ -9,14 +9,10 @@
 
 import logging
 
-from ..errors import InvalidGrantError
```
I'll update my Pycharm local settings to keep separate lines instead of group. That'll fix these import changes.

zachswasey commented on authlib/oauth2/rfc6749/models.py:
```diff
@@ -165,6 +165,16 @@ def get_scope(self):
         """
         raise NotImplementedError()
 
+    def get_dpop_jkt(self):
```
Mixin.

zachswasey commented on authlib/oauth2/rfc6749/models.py:
```diff
@@ -241,3 +251,23 @@ def get_client(self):
+        """
+        raise NotImplementedError()
+
+    def get_dpop_jkt(self):
```
Mixin.

zachswasey commented on authlib/oauth2/rfc6749/resource_protector.py:
```diff
@@ -39,6 +38,31 @@ def scope_insufficient(token_scopes, required_scopes):
 
         return True
 
+    def validate_token_type(self, token, request):
+        from ..rfc6750.errors import InvalidTokenError
+        from ..rfc9449 import DPoPTokenValidator
+        if token.get_dpop_jkt():
+            if self.TOKEN_TYPE != DPoPTokenValidator.TOKEN_TYPE:
+                raise InvalidTokenError(description=f"Access token is bound to a DPoP proof, but token type is {self.TOKEN_TYPE}",
+                                        token_type="DPoP",
+                                        realm=self.realm,
+                                        extra_attributes=self.extra_attributes)
+        else:
+            if "DPoP" in request.headers and self.TOKEN_TYPE != DPoPTokenValidator.TOKEN_TYPE:
+                raise InvalidTokenError(
+                    token_type=self.TOKEN_TYPE,
+                    description=f"DPoP proof not expected for {self.TOKEN_TYPE} token type",
+                    realm=self.realm,
+                    extra_attributes=self.extra_attributes
+                )
+
+        saved_token_type = token.get_token_type()
+        if saved_token_type.lower() != self.TOKEN_TYPE:
+            raise InvalidTokenError(description=f"Access token is of type {saved_token_type}, but token type is {self.TOKEN_TYPE}",
+                                    token_type=saved_token_type,
+                                    realm=self.realm,
+                                    extra_attributes=self.extra_attributes)
```
Mixin, if possible. Though this was a bit more difficult to do, but I'll give it another look.

zachswasey commented:
Based on the discussions on #814 I'll likely refactor the server-side component a bit, so take the existing structure with a grain of salt.

zachswasey commented:
Refactored and addressed most of the comments I left above, and from ideas from the PAR #814 PR. The one remaining bit is in `rfc6749/resource_protector.py` and `rfc6749/requests.py`, but I can address that after overall architecture comments.

I haven't fully finished the httpx/requests integrations yet, but I refactored their Auth classes to remove code duplication and make it easier to add new extension Auths in the future via a CompositeAuth, and handle retrying for auth reasons.

Again, please give feedback on the overall approach before I address cleanup, tests, docs, etc. I added an overall server extension which automatically hooks into AuthorizationCodeGrant and the RefreshTokenGrant, so developers don't have to add the same extension multiple times (like in the original top-level comment).

```
dpop_nonce_generator = DefaultDPoPNonceGenerator()
dpop_proof_validator = DPoPProofValidator(nonce_generator=dpop_nonce_generator)
server.register_extension(DPoP(proof_validator=dpop_proof_validator)

require_oauth = ResourceProtector()
require_oauth.register_token_validator(MyDPoPTokenValidator(proof_validator=dpop_proof_validator))
```

zachswasey commented on authlib/oauth2/client.py:
```diff
@@ -93,6 +96,11 @@ def __init__(
+        self.dpop_proof = dpop_proof
+        if self.dpop_proof:
+            self.dpop_proof.generate_jwk(token, metadata.get("dpop_signing_alg_values_supported", None))
+            self.dpop_auth = DPoPAuth(self.dpop_proof)
```
Add `None` above.
